src/TestFramework/Coverage/JUnit/JUnitTestCaseSorter.php

```diff
@@ -74,15 +93,54 @@ public function getUniqueSortedFileNames(array $tests): iterable
          */
 
         // sort tests to run the fastest first
-        usort(
-            $uniqueTestLocations,
-            static function (TestLocation $a, TestLocation $b) {
-                return $a->getExecutionTime() <=> $b->getExecutionTime();
+        foreach (self::sort($uniqueTestLocations) as $testLocation) {
+            /** @var string $filePath */
+            $filePath = $testLocation->getFilePath();
+
+            yield $filePath;
+        }
+    }
+
+    /**
+     * Sorts tests to run the fastest first. Exposed for benchmarking purposes.
+     *
+     * @param TestLocation[] $uniqueTestLocations
+     *
+     * @return iterable<TestLocation>
+     */
+    public static function sort(array $uniqueTestLocations): iterable
+    {
+        // Pre-sort first buckets, optimistically assuming that
+        // most projects won't have tests longer than a second
+        $buckets = [
+            0 => [],
+            1 => [],
+            2 => [],
+            3 => [],
+            4 => [],
+            5 => [],
+            6 => [],
+            7 => [],
+        ];
+
+        foreach ($uniqueTestLocations as $location) {
+            // Quicky drop off lower bits, reducing precision to 8th of a second
+            $msTime = $location->getExecutionTime() * 1024 >> 7; // * 1024 / 128
+
+            // For anything above 4 seconds reduce precision to 4 seconds
+            if ($msTime > 32) {
+                $msTime = $msTime >> 5 << 5; // 7 + 5 = 12 bits
             }
-        );
 
-        foreach ($uniqueTestLocations as $testLocation) {
-            yield $testLocation->getFilePath();
+            $buckets[$msTime][] = $location;
+        }
+
+        ksort($buckets);
+
+        foreach ($buckets as $bucket) {
+            foreach ($bucket as $value) {
```

Comment:
```suggestion
            yield from $bucket;
```

Reply:
It'll break the order  :man_facepalming:

Reply:
How do you mean?

Reply:
Wish I knew! Try it: check it out, change, run the test.

Reply:
Actually the answer is silly :) With `yield from` you also yield the keys, but hey you know what's the default value of `iterator_to_array` second argument? `true`: https://3v4l.org/XWc5G

Reply:
Well, that's silly for sure. Fixed this.